File: AOS/app/src/test/java/com/example/escapephone/ThemeProgressTests.kt
package com.example.escapephone

import com.example.escapephone.core.model.EndingType
import com.example.escapephone.core.model.GameProgress
import com.example.escapephone.core.model.GameStage
import com.example.escapephone.core.persistence.TimeProvider
import com.example.escapephone.core.theme.InMemoryThemeProgressStore
import com.example.escapephone.core.theme.ThemeId
import com.example.escapephone.core.theme.ThemeStatus
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertNull
import org.junit.Assert.assertTrue
import org.junit.Test

private class FixedTimeProvider(private val value: Long) : TimeProvider { override fun now() = value }

class ThemeProgressTests {
    @Test
    fun migrateLegacyProgress_createsTheLastCommitProgress() {
        val legacy = GameProgress(currentStage = GameStage.messengerSolved, startedAt = 100L, hintCount = 2)
        val store = InMemoryThemeProgressStore(timeProvider = FixedTimeProvider(500L))
        store.setLegacyGameProgress(legacy)
        store.migrateLegacyProgressIfNeeded()
        val migrated = store.loadThemeProgress(ThemeId.theLastCommit)
        assertEquals(ThemeId.theLastCommit, migrated.themeId)
        assertEquals(ThemeStatus.inProgress, migrated.themeStatus)
        assertEquals(2, migrated.hintCount)
        assertEquals(100L, migrated.startedAt)
    }

    @Test
    fun migrateLegacyProgress_preservesCompletedStage() {
        val legacy = GameProgress(currentStage = GameStage.gameCompleted, completedAt = 900L, endingType = EndingType.publicDisclosure, collectedEvidenceIds = setOf("root_m"))
        val store = InMemoryThemeProgressStore(timeProvider = FixedTimeProvider(1000L))
        store.setLegacyGameProgress(legacy)
        store.migrateLegacyProgressIfNeeded()
        val migrated = store.loadThemeProgress(ThemeId.theLastCommit)
        assertEquals(ThemeStatus.completed, migrated.themeStatus)
        assertEquals(900L, migrated.completedAt)
        assertEquals("publicDisclosure", migrated.endingId)
        assertTrue(migrated.collectedEvidenceIds.contains("root_m"))
    }

    @Test
    fun migrateLegacyProgress_doesNotRunTwice() {
        val legacy = GameProgress(currentStage = GameStage.messengerSolved)
        val store = InMemoryThemeProgressStore(timeProvider = FixedTimeProvider(10L))
        store.setLegacyGameProgress(legacy)
        store.migrateLegacyProgressIfNeeded()
        store.resetThemeProgress(ThemeId.theLastCommit)
        store.migrateLegacyProgressIfNeeded()
        val progress = store.loadThemeProgress(ThemeId.theLastCommit)
        assertEquals(ThemeStatus.notStarted, progress.themeStatus)
        assertNull(progress.startedAt)
    }

    @Test
    fun resetThemeProgress_doesNotResetOtherTheme() {
        val store = InMemoryThemeProgressStore()
        store.saveThemeProgress(store.loadThemeProgress(ThemeId.theLastCommit).copy(hintCount = 3))
        store.saveThemeProgress(store.loadThemeProgress(ThemeId.convenienceStoreLoop).copy(hintCount = 5))
        store.resetThemeProgress(ThemeId.theLastCommit)
        assertEquals(0, store.loadThemeProgress(ThemeId.theLastCommit).hintCount)
        assertEquals(5, store.loadThemeProgress(ThemeId.convenienceStoreLoop).hintCount)
    }

    @Test
    fun resetAllThemeProgress_clearsEveryTheme() {
        val store = InMemoryThemeProgressStore()
        store.saveThemeProgress(store.loadThemeProgress(ThemeId.theLastCommit).copy(hintCount = 3))
        store.saveThemeProgress(store.loadThemeProgress(ThemeId.convenienceStoreLoop).copy(hintCount = 5))
        store.resetAllThemeProgress()
        assertEquals(0, store.loadThemeProgress(ThemeId.theLastCommit).hintCount)
        assertEquals(0, store.loadThemeProgress(ThemeId.convenienceStoreLoop).hintCount)
    }

    @Test
    fun themeSelection_returnsAvailableThemes() {
        val themes = com.example.escapephone.core.theme.ThemeRegistry.themes
        assertEquals(2, themes.size)
        assertTrue(themes.any { it.themeId == ThemeId.theLastCommit && it.isAvailable })
    }

    @Test
    fun themeProgress_isStoredSeparately() {
        val store = InMemoryThemeProgressStore()
        store.saveThemeProgress(store.loadThemeProgress(ThemeId.theLastCommit).copy(hintCount = 1))
        assertFalse(store.loadThemeProgress(ThemeId.convenienceStoreLoop).hintCount == 1)
    }
}

File: ios/EscapePhoneTests/ThemeProgressTests.swift
import XCTest
@testable import EscapePhone

final class ThemeProgressTests: XCTestCase {
    private struct FixedDateProvider: DateProviding { let now: Date }

    func testMigrateLegacyProgress_createsTheLastCommitProgress() {
        var legacy = GameProgress(currentStage: .messengerSolved, hintCount: 2)
        legacy.startedAt = Date(timeIntervalSince1970: 100)
        let store = InMemoryThemeProgressStore(legacyGameProgress: legacy, dateProvider: FixedDateProvider(now: Date(timeIntervalSince1970: 500)))
        store.migrateLegacyProgressIfNeeded()
        let migrated = store.loadThemeProgress(themeId: .theLastCommit)
        XCTAssertEqual(migrated.themeId, .theLastCommit)
        XCTAssertEqual(migrated.themeStatus, .inProgress)
        XCTAssertEqual(migrated.hintCount, 2)
        XCTAssertEqual(migrated.startedAt, Date(timeIntervalSince1970: 100))
    }

    func testMigrateLegacyProgress_preservesCompletedStage() {
        var legacy = GameProgress(currentStage: .gameCompleted)
        legacy.completedAt = Date(timeIntervalSince1970: 900)
        legacy.endingType = .publicDisclosure
        legacy.collectedEvidenceIds = ["root_m"]
        let store = InMemoryThemeProgressStore(legacyGameProgress: legacy, dateProvider: FixedDateProvider(now: Date(timeIntervalSince1970: 1000)))
        store.migrateLegacyProgressIfNeeded()
        let migrated = store.loadThemeProgress(themeId: .theLastCommit)
        XCTAssertEqual(migrated.themeStatus, .completed)
        XCTAssertEqual(migrated.completedAt, Date(timeIntervalSince1970: 900))
        XCTAssertEqual(migrated.endingId, "publicDisclosure")
        XCTAssertTrue(migrated.collectedEvidenceIds.contains("root_m"))
    }

    func testMigrateLegacyProgress_doesNotRunTwice() {
        let legacy = GameProgress(currentStage: .messengerSolved)
        let store = InMemoryThemeProgressStore(legacyGameProgress: legacy)
        store.migrateLegacyProgressIfNeeded()
        store.resetThemeProgress(themeId: .theLastCommit)
        store.migrateLegacyProgressIfNeeded()
        let progress = store.loadThemeProgress(themeId: .theLastCommit)
        XCTAssertEqual(progress.themeStatus, .notStarted)
        XCTAssertNil(progress.startedAt)
    }

    func testResetThemeProgress_doesNotResetOtherTheme() {
        let store = InMemoryThemeProgressStore()
        var lastCommit = store.loadThemeProgress(themeId: .theLastCommit); lastCommit.hintCount = 3
        store.saveThemeProgress(lastCommit)
        var convenience = store.loadThemeProgress(themeId: .convenienceStoreLoop); convenience.hintCount = 5
        store.saveThemeProgress(convenience)
        store.resetThemeProgress(themeId: .theLastCommit)
        XCTAssertEqual(store.loadThemeProgress(themeId: .theLastCommit).hintCount, 0)
        XCTAssertEqual(store.loadThemeProgress(themeId: .convenienceStoreLoop).hintCount, 5)
    }

    func testResetAllThemeProgress_clearsEveryTheme() {
        let store = InMemoryThemeProgressStore()
        var lastCommit = store.loadThemeProgress(themeId: .theLastCommit); lastCommit.hintCount = 3
        store.saveThemeProgress(lastCommit)
        var convenience = store.loadThemeProgress(themeId: .convenienceStoreLoop); convenience.hintCount = 5
        store.saveThemeProgress(convenience)
        store.resetAllThemeProgress()
        XCTAssertEqual(store.loadThemeProgress(themeId: .theLastCommit).hintCount, 0)
        XCTAssertEqual(store.loadThemeProgress(themeId: .convenienceStoreLoop).hintCount, 0)
    }

    func testThemeSelection_returnsAvailableThemes() {
        let themes = ThemeRegistry.themes
        XCTAssertEqual(themes.count, 2)
        XCTAssertTrue(themes.contains { $0.themeId == .theLastCommit && $0.isAvailable })
    }

    func testThemeProgress_isStoredSeparately() {
        let store = InMemoryThemeProgressStore()
        var lastCommit = store.loadThemeProgress(themeId: .theLastCommit); lastCommit.hintCount = 1
        store.saveThemeProgress(lastCommit)
        XCTAssertNotEqual(store.loadThemeProgress(themeId: .convenienceStoreLoop).hintCount, 1)
    }
}
